homeassistant/components/esphome/bluetooth/client.py

```diff
@@ -317,7 +317,20 @@ def _on_bluetooth_connection_state(
                 connected_future.cancel()
                 raise
             await connected_future
-        await self.get_services(dangerous_use_bleak_cache=dangerous_use_bleak_cache)
+
+        try:
+            await self.get_services(dangerous_use_bleak_cache=dangerous_use_bleak_cache)
+        except asyncio.CancelledError:
+            # On cancel we must still raise cancelled error
+            # to avoid blocking the cancellation even if the
+            # disconnect call fails.
+            with contextlib.suppress(Exception):
+                await self.disconnect()
+            raise
+        except Exception:
```

Comment:
I realize it likely makes sense to run the disconnect cleanup on every possible error, however, might the debug/error logging change if its a `BleakError` which you expect now and want to swallow vs any other new `Exception` types that might represent coding bugs (e.g. a `TypeError`?

I see there is prior art here of catching `Exception` above, so suspect there is a motivation for this. I am not familiar with this library and the context, so I understand if i'm missing some important context here.

Reply:
I didn't want to swallow the exception here since python will tell us that the exception was directly caused by another exception if the disconnect raises here.

Reply:
I didn't want to swallow the original exception since python will already work out that there was an exception in `disconnect` while handling the original exception.

Reply:
I also missed the bare `raise` which I believe is re-raising as is